pkg/controller/secretsmanager/secret/setup.go

```diff
@@ -113,6 +119,86 @@ type hooks struct {
 	kube   client.Client
 }
 
+func (e *hooks) lateInitialize(cr *svcapitypes.SecretParameters, resp *svcsdk.DescribeSecretOutput) error {
+	payload, err := e.getPayload(context.TODO(), cr)
+	if err := client.IgnoreNotFound(err); err != nil {
+		return err
+	}
+
+	// If the K8s does not exist, create it with the data from AWS
+	if payload == nil {
+		req := &svcsdk.GetSecretValueInput{
+			SecretId: resp.ARN,
+		}
+		valueResp, err := e.client.GetSecretValueWithContext(context.TODO(), req)
+		if err != nil {
+			return errors.Wrap(err, errGetSecretValue)
+		}
+		ref, err := getSecretRef(cr)
+		if err != nil {
+			return err
+		}
+
+		sc := &corev1.Secret{}
+		sc.ObjectMeta.Name = ref.Name
+		sc.ObjectMeta.Namespace = ref.Namespace
+
+		data, err := getAWSSecretData(ref, valueResp)
+		if err != nil {
+			return errors.Wrap(err, errGetAWSSecretValue)
+		}
+		sc.Data = data
+		return errors.Wrap(e.kube.Create(context.TODO(), sc), errCreateK8sSecret)
+	}
+
+	return nil
+}
+
+func getAWSSecretData(ref *svcapitypes.SecretReference, s *svcsdk.GetSecretValueOutput) (map[string][]byte, error) { // nolint:gocyclo
+	if ref.Key != nil {
+		switch {
+		case awsclients.StringValue(s.SecretString) != "":
+			return map[string][]byte{
+				*ref.Key: []byte(awsclients.StringValue(s.SecretString)),
+			}, nil
+		case len(s.SecretBinary) != 0:
+			return map[string][]byte{
+				*ref.Key: s.SecretBinary,
+			}, nil
+		default:
+			return nil, errors.New(errNoAWSValue)
+		}
+	}
+
+	var raw []byte
+
+	switch {
+	case awsclients.StringValue(s.SecretString) != "":
+		raw = []byte(awsclients.StringValue(s.SecretString))
+	case len(s.SecretBinary) != 0:
+		raw = s.SecretBinary
+	default:
+		return nil, errors.New(errNoAWSValue)
+	}
+
+	parsed := map[string]string{}
+	err := json.Unmarshal(raw, &parsed)
+	if err != nil {
+		return nil, errors.Wrap(err, errParseSecretValue)
+	}
+
+	payload := map[string][]byte{}
+	if v, exists := parsed[awsclients.StringValue(ref.Key)]; exists {
```

Comment:
We send the data as a `string` if the key is given, I'd expect the reverse of this to happen in late-init, i.e. look for a string only if the key is given. If not, then we should always treat that the whole map is requested. So, I think we should only iterate here and not account for the case where there is a key, blob in AWS is a map and user actually meant a specific key in that whole blob in AWS.

Reply:
Oh in fact, it will never be non-nil anyway since we checked it in the beginning.